### ISSUE TITLE
scrollPositionRestoration has several problems

### ISSUE THREAD

**Issue by jnizet** [angular/angular#24547]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Performance issue
[x] Feature request
[x] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
[ ] Other... Please describe:
</code></pre>

## Current behavior

I started experimenting with the new `scrollPositionRestoration` feature in the [RouterModule extra options](https://next.angular.io/api/router/ExtraOptions). I expected scroll restoration to work by default when setting the property to 'enabled', but it doesn't. And the documentation has issues, too.

## Expected behavior

The documentation says:
> 'enabled'--set the scroll position to the stored position. This option will be the default in the future.

So I naïvely thought that setting the flag to 'enabled' would be sufficient to restore the scroll position. But it isn't. 

Indeed, the scroll event is fired, and the scroll position is restored, before the `ngAfterViewInit` hook of the activated component has been called. So the view of the component is not ready yet when the router tries to restore the scroll position (i.e. there is no way to scroll to the end of a long list, because the list isn't there yet). 

And even if it was restored after the view is ready, that would only work if the activated component used a resolved guard to load the data.

So, the documentation should, IMHO, at least indicate that restoring the scroll position always requires to
 - explicitly intercept the Scroll event, and scroll imperatively after a delay. This can be done in a single place, but I don't see how to do that in a reliable way, since there is no way to know if the delay is sufficient for the data to have been loaded (but it would at least work if resolve guards are used consistently), or
 - explicitly intercept the Scroll event in each routed component, and imperatively scroll when the data has been loaded and the view has been rendered. This is not a trivial task.

I read the remaining of the documentation, which has examples about doing this kind of stuff (although it doesn't really say that they're required). But those examples are all incorrect.

Here's the first example:

```
    class AppModule {
     constructor(router: Router, viewportScroller: ViewportScroller, store: Store<AppState>) {
       router.events.pipe(filter(e => e instanceof Scroll), switchMap(e => {
         return store.pipe(first(), timeout(200), map(() => e));
       }).subscribe(e => {
         if (e.position) {
           viewportScroller.scrollToPosition(e.position);
         } else if (e.anchor) {
           viewportScroller.scrollToAnchor(e.anchor);
         } else {
           viewportScroller.scrollToPosition([0, 0]);
         }
       });
     }
    }
```

This example uses a Store service, which is not part of Angular (I guess it's part of ngrx). So that makes it hard to understand and adapt for those who don't use ngrx.

Besides, it doesn't compile, because a closing parenthesis is missing, and because `e` is of type Event, and not of type Scroll, and thus has no `position` property. 

The second example is the following:
```
    class ListComponent {
      list: any[];
      constructor(router: Router, viewportScroller: ViewportScroller, fetcher: ListFetcher) {
        const scrollEvents = router.events.filter(e => e instanceof Scroll);
        listFetcher.fetch().pipe(withLatestFrom(scrollEvents)).subscribe(([list, e]) => {
          this.list = list;
          if (e.position) {
            viewportScroller.scrollToPosition(e.position);
          } else {
            viewportScroller.scrollToPosition([0, 0]);
          }
        });
      }
    }
```

It doesn't compile because it still uses an old, non-pipeable operator, and because, once again, `e` is of type Event, not Scroll.

But even after fixing the compilation errors, it doesn't work because the view hasn't been updated with the new list yet when `viewportScroller.scrollToPosition(e.position);` is called. 

So the code would have to be changed to the following in order to compile and work as expected

```
    class ListComponent {
      list: any[];
      constructor(router: Router, viewportScroller: ViewportScroller, fetcher: ListFetcher) {
        const scrollEvents = router.events.filter(e => e instanceof Scroll);
        listFetcher.fetch().pipe(withLatestFrom(scrollEvents)).subscribe(([list, e]) => {
          this.races = list;
          const scrollEvent = e as Scroll;
          of(scrollEvent).pipe(delay(1)).subscribe(s => {
            if (s.position) {
              viewportScroller.scrollToPosition(s.position);
            } else {
              viewportScroller.scrollToPosition([0, 0]);
            }
          });
        });
      }
    }
```

I think that none of these solutions is really simple enough, though. Here are two ideas that could maybe make things easier:

 - only fire the Scroll event and try to restore the position after the ngAfterViewInit hook has been called. This should at least make things work when a resolve guard is used to load the list. Or when the list is available immediately.
 - for the other cases, allow to inject a service that the component could call when the list has been loaded. It would be up to this service to get the last scroll position or anchor, to wait until the view has been rendered, and then to restore the scroll position. It would ignore all but the first call after the component has been activated.

## Minimal reproduction of the problem with instructions

Here's a repo illustrating the various issues and solutions presented above: https://github.com/jnizet/scrollbug. It's a standard angular-cli project. I can't run it in stackblitz unfortunately (probably because Stackblitz doesn't support the beta release of Angular).

## What is the motivation / use case for changing the behavior?

First, the documentation should be fixed and made clearer
1. it should not use ngrx
2. it should contain examples that compile, and run as expected
3. it should make it clear than simply setting the flag to 'enabled' is not sufficient to enable scroll restoration 

Second, it should be way easier to make that feature work. See ideas above.

## Environment

<pre><code>
Angular version: 6.1.0-beta.1

Browser:
- [x] Chrome (desktop) version 67.0.3396.87
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [x] Firefox version 60.0.2 
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
</code></pre>

**Comment by jasonaden** [comment]
Thanks for the detailed issue report! Great to get feedback on this new feature and hopefully get these things addressed quickly.

**Comment by mhevery** [comment]
Closed by https://github.com/angular/angular/pull/20030

**Comment by jnizet** [comment]
@mhevery @jasonaden please reopen. 

This issue is not about a missing scroll position restoration. It's precisely about the new scroll position restoration feature that was introduced in 6.1.0-beta.1. 

And it thus can't be closed by #20030: this issue is precisely about several problems in the code and documentation *introduced* by #20030.

**Comment by mhevery** [comment]
Sorry

**Comment by damienwebdev** [comment]
Just chiming in here, my initial expectation of this feature was that the "scrollPosition" was going to be handled internally by angular with "sane" defaults that are simple to configure ('enabled'|'disabled'|'top' make perfect sense as outlined by @vsavkin). 

But, just like @jnizet after upgrading a simple project to `6.1.0-rc1` it appears that these options don't actually do anything (maybe I'm missing something)? 

I'm curious about the following:

1. Are we expected to implement this scroll functionality on our own with the help of the `scrollPositionRestorationService`?
2. What do the flags actually do out-of-the-box?

**Comment by damienwebdev** [comment]
Just leaving this here, as I did some debugging to determine why this wasn't working... For anyone who sees this in the future. If you have the following code in your stylesheet...

```
html,body {height: 100%}
```

See here as a reference: https://stackoverflow.com/questions/1174863/javascript-scrollto-method-does-nothing

This will be a bug for you. Removing that snippet causes makes this PR do what was intended.

**Comment by jasonaden** [comment]
@damienwebdev Thanks for that note on the body height. That very well could help quite a number of people

On the other things discussed in this issue, I've talked with @vsavkin about it and we're going to solve this in a couple ways. First is some documentation to make sure your use cases are discussed and solutions given for some of the common patterns. That documentation will be forthcoming.

Another is to fix an issue that was introduced some time ago where it's possible to have a `NavigationEnd` event fire before change detection. We need to change this so `NavigationEnd` happens after CD to make sure the page is actually rendered before trying to scroll.

**Comment by StefanRein** [comment]
@jasonaden Could you tell, if following use case will be covered:

We are using **perceived performance** and intentionally are loading the data every time on component initialization, instead of caching, thus the app could be used with multiple devices at the same time.
Will there be a solution to tell the scroll restoration when my data of my component is loaded?

Thank you in forward.

**Comment by sorakthunly** [comment]
How's the progress on this issue?

**Comment by sorakthunly** [comment]
What's the progress on this issue?

**Comment by ericmartinezr** [comment]
>explicitly intercept the Scroll event in each routed component, and imperatively scroll when the data has been loaded and the view has been rendered. This is not a trivial task.

This is a must after watching Jason's video about ng6.1 features where he says : _[...] be aware on when and where you're rendering out your position_.

Video: https://youtu.be/jNsbB8V9u0A?t=6m10s

**Comment by StefanRein** [comment]
@mhevery 

I know one could preload all the data with **Resolve** from '@angular/router'. This would be fine and we could provide for each component an equal named class which could resolve all the needed data.
Then the scroll position restoration could work just as expected.
Usually this would also mean to show an overall spinner on the page, so the user would see there is something going on.

Another also very common practice is to use perceived performance. This is the way we decided to interact with preloading data. In our application we have a lot of times a list, which would show some grey boxes unless the data arrives and the list is shown.

As the component already is constructed and all the events are fired this would mean we would need to have the possibility to provide some asynchronous way to tell the data has been loaded for the scroll restoration service.

Do you see the possibility to provide a:

**1. CustomScrollPositionRestorationStrategy** or
**2.** another **Lifecycle Hook**
```javascript
interface AfterContentReady {
    ngAfterContentReady(): Promise<boolean>;
}
```

I am also open for other ways. Maybe one would have some suggestions?

**Comment by akash-pal** [comment]
Whats the status on the issue? Hash anyone found a perfect solution/ when will this issue be fixed?

**Comment by burredis** [comment]
I **used scrollIntoView(false)** 

`<router-outlet (activate)='onActivate($event)'></router-outlet>`

`onActivate(e) {
    document.getElementsByClassName('user-notification')[0].scrollIntoView(false);
  }`

`('user-notification')[0]` is the element that I used to scroll top. `window.scrollTo(0,0)`, `document.body.scrollTo(0,0)` didn't work. I found this way to workaround this issue.

I hope help :)

**Comment by StefanRein** [comment]
One additional note I would like to mention here:
In some frameworks like IONIC 3 or other applications, the scrolling element is by default not the document body.
##### (I did workaround this by setting all parent elements until the body to position: static / relative (they were position: absolute before), which introduced other problems, e.g. with modal dialogs or backdrops.)

**Feature request:**
- A possibility to register another scrolling element.
- Handle multiple elements for scrolling.

**Potential problems**
- Multiple / nested router outlets
  - Possibility to set and get the scroll position for each one 
  (**Note:** The default scrolling element could be the parent element of the router outlet element, but we would maybe need to register another one.)
- Async between different router outlets and their scrolling position until the components are finished with loading and building the desired content (perceived performance strikes here again). 
  - **Note:** Yes one could also resolve first the data. But how to handle perceived performance here? Maybe setting a component into a specific position while resolving the data?

**Comment by rankitbishnoi** [comment]
[https://github.com/angular/universal/issues/830#issuecomment-414239486](url)

**Comment by twoheaded** [comment]
It's not quite clear, is there any way to set the scroll policy for certain routes?

**Comment by jasonaden** [comment]
Thanks for the feedback so far on this issue. We're going to be making some minor improvements to the current implementation, but also will be adding some new documentation in the next few weeks to help with some of the scenarios you guys have been having trouble with. Thanks for the patience on this one.

**Comment by lnaie** [comment]
trying to use this "feature" right now. haven't seen anything lately as total bs like this one. thx so much angular team!

**Comment by lnaie** [comment]
i see you don't like my non constructive comments. sure, np!
here i have a little follow up, for v6.1.7:

scrollPositionRestoration: 'enabled'
- this value doesn't work as expected, it takes me always to the top of the page.

scrollPositionRestoration: 'disabled'
- this one makes the feature work, o.O well sort of, because only sometimes it takes me to the same scroll position.

no further comments.

**Comment by lnaie** [comment]
new insights...
looks like the feature is working when using the browser's Back button, so my apologies.

but when i use something like this, it doesn't:
`this.router.navigate(['../'], { relativeTo: this.route });`

could somebody explain why the above code is not doing the same as Location.back(), and what are the options?

**Comment by sarunint** [comment]
@lnaie That is intended. This feature will work when use browser's navigating back. Redirecting to `../` is not navigating back, it's going up one level.

**Comment by lnaie** [comment]
that explains it, thanks!

how do i know if i landed straight into a child route? 
all that to be able to use, imo, `Location.back()` or `this.router.navigate(['../'], { relativeTo: this.route });` to go back one level in the router.

**Comment by lnaie** [comment]
yes. thanks. it works with the Location.back() only. now i'm trying to selectively use that or the other way, as described above. maybe you can advise. :)

**Comment by jasonaden** [comment]
@lnaie Your comments from 3 days ago are inappropriate and violate the [Code of Conduct](https://github.com/angular/angular/blob/master/CODE_OF_CONDUCT.md). I suggest you review that, and try to communicate with professionalism and respect.

**Comment by rubenheymans** [comment]
what's the status? I just upgraded from angular 6 to 6.1 (took me 3 hours :D) to implement this feature but now it turns out it's not working at all (at least for me)
I just want the page to start at the top without writing tons of code, so I use ` scrollPositionRestoration: 'top',` ?

**Comment by jfey** [comment]
Dear Angular team, Its good so see that you are aware of this issue and working on it. Can you give as an heads-up regarding the roadmap? In the essence of it all: Would it make sense trying to fix/solve inside of our app or should be just stay put, working on other topics in the meantime till you're done.

**Comment by KroanNL** [comment]
Just to confirm what is posted above. Simple examples work like the should, but as soon as you do some kind of loading of data with a resolver, the timing will be off and the NavigationEnd-event (and therefor Scroll-event) will fire before the actual page is shown. This results in no scrolling at all.

An update on the timeline would be appreciated :)

**Comment by KroanNL** [comment]
As an addition, I want to report that when using the following code

```
        this.router.navigate([], {
            relativeTo: this.route,
            queryParams: queryParams,
            replaceUrl: true
        });
```

The routerevents, including scrolling, will happen as well. This is not expected behaviour, if you ask me. When replacing only the url (for example by changing the queryParams, as in the example above), you don't want the page to suddenly scroll to the top of the page as well.

I tried to use `skipLocationChange` as well, without success.

**Comment by burredis** [comment]
> I **used scrollIntoView(false)**
> 
> `<router-outlet (activate)='onActivate($event)'></router-outlet>`
> 
> `onActivate(e) { document.getElementsByClassName('user-notification')[0].scrollIntoView(false); }`
> 
> `('user-notification')[0]` is the element that I used to scroll top. `window.scrollTo(0,0)`, `document.body.scrollTo(0,0)` didn't work. I found this way to workaround this issue.
> 
> I hope help :)

New hope :P

@ViewChild('wrapper') wrapper: ElementRef;
onActivate(e) {
    this.wrapper.nativeElement.scrollIntoView({block: 'start', inline: 'nearest'});
 }

in my case, this block worked fine `scrollIntoView({block: 'start', inline: 'nearest'})`

**Comment by lnaie** [comment]
> @lnaie Your comments from 3 days ago are inappropriate and violate the [Code of Conduct](https://github.com/angular/angular/blob/master/CODE_OF_CONDUCT.md). I suggest you review that, and try to communicate with professionalism and respect.

done

**Comment by NikolaJankovic** [comment]
scrollPositionRestoration has another issue. If you load data using a 'load more'-type component that updates the query parameters, the default behaviour of scrolling to the top of the page breaks the default ui (additional data is rendered below the current and the scroll position should not change). That's a very common ui that should be accounted for.

There could be an additional argument that specifies the behaviour when a query parameter is updated, rather than just treating it the same as a full route change.

I don't know to sound too critical of the feature, it seems to work great on static pages, but the current solution seems like an oversimplification the second you start loading content dynamically.

**Comment by goat67** [comment]
Will this functionality support lazy loaded modules? I could not get this to work I am running 6.1.10 and Material 2.0

I had a simple requirement return to the same scroll position from a view with a list of items after viewing the details of one of the items.
 
I finally had to resort to reading and storing the scroll position in the method I used to navigate to the detail view 
const top = document.querySelector('mat-sidenav-content').scrollTop;
this.store.dispatch(new routePositionActions.SetPosition({route: 'Items', top: top, left: 0}));

I then set this stored position in the ngAfterViewInit method
document.querySelector('mat-sidenav-content').scrollTop = data.routePosition.top;

**Comment by Splaktar** [comment]
@goat67 it's a known issue with how this feature works with `md-sidenav-container` and/or other containers like Ionic uses. This is mentioned above and in https://github.com/angular/material2/issues/4280.

**Comment by mberezinski** [comment]
Any news on the topic?

**Comment by Lukasz196** [comment]
For my project in Angular 6.1.9 work solution described in link bellow (for jQuery but this is only css).
[https://stackoverflow.com/a/18573599/4490997](https://stackoverflow.com/a/18573599/4490997)

```css
html {
  height: 100%;
  overflow: auto;
} 
body {
  height: 100%;
}
```
then 

``` javascript
{scrollPositionRestoration: 'top'}
```
works

**Comment by CelticParser** [comment]
**Angular 7.1.2 | Material 7.1.1 (using 'mat-sidenav')**. Having the same issue. However,

- [Lukasz196](https://github.com/angular/angular/issues/24547#issuecomment-444598870) css solution doesn't seem to apply. 

'@HostListener('window:scroll')' functionality doesn't seem to work as well with this layout type.

@damienwebdev - Can you supply a stackblitz showing your method to how it works for you?

> I have a working app with material (which uses mat-sidenav) and all window.scroll variants work as expected.

 Side Note: _I just started a huge project three months ago using Angular for the first time and I am very disappointed that a simple scrolling function(s) after 7 version releases is still broke. We just started mocking with large data sets and realized that the UI cannot scroll to the top when the user switches views. Imagine scrolling on a mobile device dozens of items down then needing to go back a view that has just as many and having to manually thumb-it back to the top? We can't even add a simple "scroll-to-top" fab because the 'window:scroll' seems to break with Material2 as well (I think the two are related).  This is a very basic, 101 function for UX._

Should we (angular/angular) be communicating/working together with the crew at angular/material2 with this bug/issue?

**Comment by maenthoven** [comment]
Is there any ETA for this being fixed? This has been an annoyance for my users :(

**Comment by havran** [comment]
It's possible disable anchorScrolling for some routes? (For example for auth callback rout, which use #access_token=... Thanks.

**Comment by Bretto** [comment]
I have been trying this feature, this is what I found: 

- It doesn't work with dynamic content.
- It doesn't work with transition. 
- Currently doing a simple master/detail of scrolling cards (like Instagram) with transition back and forth is just impossible without hacking it all.  

> This is a very basic, 101 function for UX

 We need a simple and elegant solution to this to stay competitive IMHO.

**Comment by JasonAlanKennedy** [comment]
Hate to say it but I left for VueJS over this. This was the straw that broke the camel's back. So many basic functions in Angular that's been open issues for a year+

**Comment by harm-less** [comment]
I too struggled with this issue but I managed to fix it with a polyfill from Ben Nadel!

Links to all important sources and explanations are in his blog post about the polyfill:
https://www.bennadel.com/blog/3534-restoring-and-resetting-the-scroll-position-using-the-navigationstart-event-in-angular-7-0-4.htm

You can check out the demo over here:
http://bennadel.github.io/JavaScript-Demos/demos/router-retain-scroll-polyfill-angular7/with-polyfill/#/app

For example: Go to "Section B", wait for the content to load and scroll a bit > go to "Section C" > press back button > profit!

It works like a charm!
The only thing that might be handy is making this an NPM package as I've copied the code into my project.

P.s.: I'm using Angular Material and even with the `mat-sidenav` component I managed to get this to work as expected 👍

**Comment by JasonAlanKennedy** [comment]
Oh yes, I know you can do that but it boils down to UX - Do I want the user have to download more and create more over head on the browser?

**Comment by harm-less** [comment]
@JasonAlanKennedy Can't argue with that, but that does make me wonder how Vue fixed this issue. Because they too must have had the same problems and had to write code for it.

**Comment by JasonAlanKennedy** [comment]
I wonder too - especially when the core is literally half the weight (and caries more out of the box!).

**Comment by KroanNL** [comment]
Would appreciate it if you discuss other frameworks in private messages and keep the discussion ontopic. Thank you.

**Comment by NagRock** [comment]
Is there any information when this will be fixed? It is strange that we got to change `onSameUrlNavigation` to get `scrollPositionRestoration` working. 👎

**Comment by tuurbo** [comment]
Until angular addresses the issue, this solution worked best for me. You might need to tweak for your use case. I borrowed some of the code from this blog post https://blog.angularindepth.com/reactive-scroll-position-restoration-with-rxjs-792577f842c

```
import { Event, Routes, RouterModule, Route, Router, NavigationStart, NavigationEnd } from '@angular/router';

interface ScrollPositionRestore {
    event: Event;
    positions: { [K: number]: number };
    trigger: 'imperative' | 'popstate' | 'hashchange';
    idToRestore: number;
}

@NgModule(...)
export class AppRoutingModule {

    constructor(
        public router: Router,
    ) {
        let lastScrollTop = 0;

        fromEvent(window, 'scroll')
            .pipe(
                debounceTime(50),
                tap(() => {
                    lastScrollTop = document.querySelector('html').scrollTop;
                }),
            )
            .subscribe();

        this.router.events
            .pipe(
                filter(
                    event =>
                        event instanceof NavigationStart || event instanceof NavigationEnd,
                ),
                scan<Event, ScrollPositionRestore>((acc, event) => {
                    return {
                        event,
                        positions: {
                            ...acc.positions,
                            ...(event instanceof NavigationStart
                                ? {
                                    [event.id]: lastScrollTop,
                                } : {}),
                        },
                        trigger:
                            event instanceof NavigationStart
                                ? event.navigationTrigger
                                : acc.trigger,
                        idToRestore:
                            (event instanceof NavigationStart &&
                                event.restoredState &&
                                event.restoredState.navigationId + 1) ||
                            acc.idToRestore,
                    };
                }),
                filter(({ event, trigger }) => event instanceof NavigationEnd && !!trigger),
                observeOn(asyncScheduler),
            )
            .subscribe(({ trigger, positions, idToRestore }) => {
                if (trigger === 'imperative') {
                    document.querySelector('html').scrollTop = 0;
                }

                if (trigger === 'popstate') {
                    document.querySelector('html').scrollTop = positions[idToRestore];
                }
            });
    }
}
```

**Comment by Boooober** [comment]
@StefanRein, thanks for your example, it worked for me.
But I don't like to put any code inside the constructor, except injections. And if we move it to the ngOnInit, it won't work.

So, inspired by your solution, I desided to move subscription to the root App component and listen router event there. And in child component we only need to trigger scrollToPosition from the root. It is more reusable and no code inside constructor :)

```js
export class AppComponent implements OnInit, OnDestroy {

    private scrollPosition: [number, number] = [0, 0];
    private scrollPositionSubscription: Subscription = new Subscription();

    constructor(
        private router: Router,
        private viewportScroller: ViewportScroller
    ) {}

    ngOnInit(): void {
        this.scrollPositionSubscription = this.router.events.pipe(
            filter((e: Event) => e instanceof Scroll),
            tap((e: Event) => {
                this.scrollPosition = (e as Scroll).position ? (e as Scroll).position : [0, 0])
            }
        ).subscribe();
    }

    ngOnDestroy(): void {
        this.scrollPositionSubscription.unsubscribe();
    }

    restoreScrollPosition(): void {
        this.viewportScroller.scrollToPosition(this.scrollPosition);
    }
}
```

...and child component...

```js
export class McModeFeedComponent {
    constructor(
        @Inject(forwardRef(() => AppComponent)) private appComponent: AppComponent
    ) {}

    ngAfterViewInit(): void {
        this.appComponent.restoreScrollPosition();
    }
}
```

**Comment by lynxSven** [comment]
@Boooober 

Workaround worked for me.
Just one note.
Your code has one ")" to much in the end.

> ```js
>                 this.scrollPosition = (e as Scroll).position ? (e as Scroll).position : [0, 0])
> ```

Could you change it in your suggestion?

**Comment by simeyla** [comment]
#### Using mat-dialog with angular material and having issues?
You must wait for the dialog to close before redirecting.

      this.dialogRef.afterClosed().subscribe(() => {
          this.router.navigate(redirect);
      }); 

Note that `afterClosed()` emits only once and doesn't need unsubscribing from.

**Comment by Boooober** [comment]
@lynxSven Thanks for pointing out this issue. Fixed

**Comment by Splaktar** [comment]
An Angular Material issue that is more specific to this discussion has been opened in https://github.com/angular/material2/issues/14996.

**Comment by aguilarguisado** [comment]
> On Chrome, Using min-height doesn't break auto-scrolling and will force body t be at least screen height.
> https://stackoverflow.com/questions/3740722/min-height-does-not-work-with-body
> 
> ```
> html {
>     min-height: 100%;
>     display: flex;
> }
> body {
>     min-height: 100%;
>     flex: 1;
> }
> ```

Thank you very much @ouijan . After having searched several hours this was the fix. The body in my project had height:100%. After changin to min-height everything works well.  One or two beers for you.

**Comment by ArtDesire** [comment]
is this ever gonna be fixed?

**Comment by JustDoItSascha** [comment]
I have the same problems. None of the solutions above worked. I have no height: 100% on my html or body. I use Angular 8 + Resolver. So the routing just ends when the data is loaded and the view is ready. But nevertheless, it doesn't work! On every route and every back button it's scrolling to top.

When I remove the option completely, then the scrolling to the top doesn't work, too.

**Comment by rickvandermey** [comment]
When using with rxjs, all i needed to do, was adding a delay of zero. All is working smoothly, except IE. Any chance an scroll polyfill needs to be added?

app.component.ts
```
constructor(
	private cd: ChangeDetectorRef,
	private readonly router: Router,
	public swUpdate: SwUpdate,
	private viewportScroller: ViewportScroller,
) {
       this.router.events
		.pipe(
			filter((e: any): e is Scroll => e instanceof Scroll),
			delay(0),
		)
		.subscribe(e => {
			if (e.position) {
				viewportScroller.scrollToPosition(e.position);
			} else if (e.anchor) {
				viewportScroller.scrollToAnchor(e.anchor);
			} else {
				viewportScroller.scrollToPosition([0, 0]);
			}
		});
}
```
app-routing.module.ts
```
RouterModule.forRoot(appRoutes, {
	enableTracing: false,
	onSameUrlNavigation: 'reload',
	scrollPositionRestoration: 'enabled',
}),
```

**Comment by nFangueiro** [comment]
> @damienwebdev Thanks for that note on the body height. That very well could help quite a number of people
> 
> On the other things discussed in this issue, I've talked with @vsavkin about it and we're going to solve this in a couple ways. First is some documentation to make sure your use cases are discussed and solutions given for some of the common patterns. That documentation will be forthcoming.
> 
> Another is to fix an issue that was introduced some time ago where it's possible to have a `NavigationEnd` event fire before change detection. We need to change this so `NavigationEnd` happens after CD to make sure the page is actually rendered before trying to scroll.

I actually need html, body height="100%" in order to use cdk-virtual-scroll-viewport with height="100%".  This makes scrollPositionRestoration incompatible with cdk-virtual-scroll...

Any advice? thanks.

**Comment by nFangueiro** [comment]
> Just leaving this here for all future visitors... as I did some debugging to determine why this wasn't working... For anyone who sees this in the future. If you have the following code in your stylesheet...
> 
> ```
> html,body {height: 100%}
> ```
> 
> [This will be a bug for you. ](https://stackoverflow.com/questions/1174863/javascript-scrollto-method-does-nothing.)
> 
> Removing that fixes that and makes this PR do what was intended. Granted, removing that style may make a bunch of other things break, but that's a different problem.
> 
> **Ninja edit:**
> It also appears that this seems to intermittently work with https://github.com/angular/material2
> 
> _Working Version (My own personal anecdote):_
> I have a working app with material (which uses `mat-sidenav`) and all `window.scroll` variants work as expected.
> 
> _Non-working version (as reported by @crisbeto ):_
> [angular/components#11552](https://github.com/angular/components/issues/11552)

I actually need html, body height="100%" in order to use cdk-virtual-scroll-viewport with height="100%". This makes scrollPositionRestoration incompatible with cdk-virtual-scroll...

Any advice? thanks.

**Comment by sergiojoker11** [comment]
Any news on this front? 
Is this @ouijan 's [workaround](https://github.com/angular/angular/issues/24547#issuecomment-419282372) safe?
Thanks

**Comment by ArtDesire** [comment]
made a simple stackblitz to demonstrate the bug and how 
https://stackblitz.com/edit/angular-jjsm86
Steps to reproduce:
1. Scroll until you see the link Page.
2. click on it.
3. go back.

To 'fix' it, uncomment the line with a bunch of breaks in page.component.html file and try again.

P.S. I use breaks as oppose to height property, so make sure there's enough of them for scroll to appear.

**Comment by Fattouh92** [comment]
@Boooober @StefanRein I tried to follow your approach but I have a doubt about it, listening to router scroll events can result that user scrolls on page A then goes to page B, scrolls again on page B, then on restoring position when going back to page A you lose the information since user scrolled on page B. Shouldn't scroll position be tied to each page?

I also cannot get router scroll events to fire properly and I do not have `height: 100%` in the app. any idea what I need to configure so the scroll events fire?

**Comment by MartinJHItInstituttet** [comment]
@aguilarguisado This worked for me.

**Comment by nmaddirala** [comment]
I'm using Angular 8, and `scrollPositionRestoration: 'enabled` does not work. I've also tried to listen to NavigationEnd events and set Windows to (0,0), and that failed. Solutions by @rickvandermey, @Boooober, @tuurbo did not work for me. 

Wondering if anyone has tried this: https://auralinna.blog/post/2018/scroll-to-top-on-angular-route-change

**Comment by alexgaspard** [comment]
Thanks to @tuurbo answer, I had it work by just adding one line to [Angular example](https://angular.io/api/router/ExtraOptions#scrollPositionRestoration): 
```
 constructor(router: Router, viewportScroller: ViewportScroller) {
    router.events.pipe(
      filter((e: Event): e is Scroll => e instanceof Scroll),
      observeOn(asyncScheduler)
    ).subscribe(e => {
      if (e.position) {
        // backward navigation
        viewportScroller.scrollToPosition(e.position);
      } else if (e.anchor) {
        // anchor navigation
        viewportScroller.scrollToAnchor(e.anchor);
      } else {
        // forward navigation
        viewportScroller.scrollToPosition([0, 0]);
      }
    });
  }
```

**Comment by Denis-Evseev** [comment]
Tried everything mentioned here - still doesn't work...

**Comment by pacocom** [comment]
I have the same problems. I'm using Angular 8, and `scrollPositionRestoration: 'enabled'` don't work when I open a routable angular material bottomsheet full screen.
When I open the angular material bottomsheet (from bottom to top in full screen) 1ms before appears the bottomsheet the current view scroll top.

Any help!
Thank!

**Comment by parliament718** [comment]
I'm having this issue too where simply doesnt work.

It starts working only when I remove ```position:absolute;``` on an element which stretches to fit the screen via top, bottom, left, right all set to 0. 

In my case the issue is when using https://www.npmjs.com/package/ng-sidebar  as it uses an element like that (stretching to fit screen via position: absolute).

Unfortunately removing position: absolute seems to cause a myriad of other issues, some very difficult one related to my usage of sticky elements via position: sticky.

Can ```scrollPositionRestoration``` please be fixed to work in this case? I'm sure many are experiencing this issue due to using ```position: absolute``` being present at some level in their DOM of container elements, perhaps without realizing it. 

In some cases like mine it's very difficult to workaround it.

**Comment by mattcwebster** [comment]
is there any date on resolution for this.  Just discovered today on my app and is depressing to know the the scrollRestoration: enabled property doesn't work as expected.  For me it works on mobile though which is odd.

**Comment by Splaktar** [comment]
@mattcwebster no, there is no ETA on a resolution. Please 👍 the first post to express your desire to have it fixed.

**Comment by parliament718** [comment]
@Splaktar 
This issue appears 15th out of 2,800 other open issues when sorted by 👍.
I'd like to hope that makes it a sort of priority.

**Comment by Splaktar** [comment]
Agreed. I will continue to lobby the team for it. Hopefully this will get into a v9 minor or v10. 🤞🏼

**Comment by mattcwebster** [comment]
@Splaktar I gave it the .  IS there any Clean work around... this is like a major thing to have on a site with a lot of content on the homepage that keeps growing.

**Comment by mattcwebster** [comment]
> When using with rxjs, all i needed to do, was adding a delay of zero. All is working smoothly, except IE. Any chance an scroll polyfill needs to be added?
> 
> app.component.ts
> 
> ```
> constructor(
> 	private cd: ChangeDetectorRef,
> 	private readonly router: Router,
> 	public swUpdate: SwUpdate,
> 	private viewportScroller: ViewportScroller,
> ) {
>        this.router.events
> 		.pipe(
> 			filter((e: any): e is Scroll => e instanceof Scroll),
> 			delay(0),
> 		)
> 		.subscribe(e => {
> 			if (e.position) {
> 				viewportScroller.scrollToPosition(e.position);
> 			} else if (e.anchor) {
> 				viewportScroller.scrollToAnchor(e.anchor);
> 			} else {
> 				viewportScroller.scrollToPosition([0, 0]);
> 			}
> 		});
> }
> ```
> 
> app-routing.module.ts
> 
> ```
> RouterModule.forRoot(appRoutes, {
> 	enableTracing: false,
> 	onSameUrlNavigation: 'reload',
> 	scrollPositionRestoration: 'enabled',
> }),
> ```

@rickvandermey thanks your solution worked for me

**Comment by mattcwebster** [comment]
For me this was the most elegant solution to track scroll position on elements other than html / body.

app.component.ts
`{
    // subscribes to router events to mark when navigation backward happened
    this.router.events
    .pipe(
      filter(x => x instanceof NavigationStart),
      tap((x: NavigationStart) => {
        if (x.navigationTrigger === 'popstate') {
          this.windowService.wasPopState = true;
        } else {
          this.windowService.wasPopState = false;
        }
      })
    ).subscribe();
   }`

Then make a directive you can attach to element you wish to track scroll.  
`import { Directive, HostListener, ElementRef } from '@angular/core';
import { WindowService } from '../services/window.service';

@Directive({
  selector: '[track-scroll]',
})

export class TrackScrollDirective {
  constructor(
    private windowService: WindowService,
    ) { }

  @HostListener('scroll', ['$event'])
  scrollIt() { 
    this.windowService.homeScroll = (event.srcElement as HTMLElement).scrollTop;
  }

}`

use a service with properties for whether a popstate action occurred, and scroll postion of the page/element you wish to track

**Comment by aFarkas** [comment]
Sorry this is a mess. Here are multiple issues in one. You don't even care to split those and work them out one by one. 

Then I look into your code and see things... It really looks like code by Java developer very simple but without any further knowledge of DOM APIs and the needs of web developers.

**Comment by Splaktar** [comment]
@aFarkas I'm not really sure who you are responding to. However, please note that the [Code of Conduct](https://github.com/angular/angular/blob/master/CODE_OF_CONDUCT.md) prohibits personal attacks and insults.

**Comment by aFarkas** [comment]
@Splaktar 
Sorry if I hurt your feelings but I did not want to insult. I'm just surprised by the ignorance towards this issue and its triage.

For example there is the problem that not only the window is a scrollable element. This could be separated into a new issue and easily fixed by just adding a callback option `getScrollingElement` for the router setup, which can be utilized by the viewportscroller.

The default value of this option could look like this:

```
() => document.scrollingElement;
``` 

And people would be able to override it with what they want:

```
() => document.querySelector('.page-wrapper');
```

**Comment by Splaktar** [comment]
@aFarkas you didn't hurt my feelings. I was just trying to help you avoid getting banned. 

If you feel like you can break this down into a separate issue with a clear reproduction, please open a new issue and reference this one. 

If you feel like you have a solution to that issue, please open a PR.

> surprised by the ignorance towards this issue and its triage

It seems like there is a bit of a language barrier here. I understand that it is frustrating when an issue that is important to you is not fixed quickly. However, the team has been focused on other priorities like shipping Ivy, improving template type checking, runtime performance, etc. Hopefully some of the feature areas that have gotten less attention lately (forms, router) will get some attention post-version 9 (after Ivy ships).

**Comment by iank-** [comment]
Hi, is there an ETA on this fix?

**Comment by Splaktar** [comment]
@iank- not at this time. An update on this was just posted 12 days ago. If there are more updates to provide, they will be posted here. It's not beneficial to ask periodically for an update as that doesn't provide value to the 52 other subscribers to this issue.

If you have a critical use case that you need to have addressed and want to raise the priority of issues like this, you can reach out to devrel@angular.io.

**Comment by PapaNappa** [comment]
Another issue I would like to point out is that scroll restoration does not work well with router animations.
If you navigate to a new route, and thus trigger a new animation, both components are displayed at the same time (which is fine, because you want a smooth transition).
However, the 'scroll to top' behaviour for non-back navigation kicks in right at the start of the animation, meaning that the previous component is scrolling to the top. This makes the animations look less smooth.
If we have a sane and sound way to control when scrolling should happen, this issue could potentially also be solved.

**Comment by AbdulKhader4123** [comment]
> When using with rxjs, all i needed to do, was adding a delay of zero. All is working smoothly, except IE. Any chance an scroll polyfill needs to be added?
> 
> app.component.ts
> 
> ```




> constructor(
> 	private cd: ChangeDetectorRef,
> 	private readonly router: Router,
> 	public swUpdate: SwUpdate,
> 	private viewportScroller: ViewportScroller,
> ) {
>        this.router.events
> 		.pipe(
> 			filter((e: any): e is Scroll => e instanceof Scroll),
> 			delay(0),
> 		)
> 		.subscribe(e => {
> 			if (e.position) {
> 				viewportScroller.scrollToPosition(e.position);
> 			} else if (e.anchor) {
> 				viewportScroller.scrollToAnchor(e.anchor);
> 			} else {
> 				viewportScroller.scrollToPosition([0, 0]);
> 			}
> 		});
> }
> ```
> 
> app-routing.module.ts
> 
> ```
> RouterModule.forRoot(appRoutes, {
> 	enableTracing: false,
> 	onSameUrlNavigation: 'reload',
> 	scrollPositionRestoration: 'enabled',
> }),
> ```

it worked for me @rickvandermey

**Comment by dsebastien** [comment]
I've tried using scrollPositionRestoration and indeed face the troubles mentioned in this issue. I've documented my findings and current solution here, just in case anyone is also looking for a solution: https://medium.com/@dSebastien/handling-scrolling-on-angular-router-transitions-e7652e57d964

It's far from perfect, but it's an improvement for apps that have fixed elements.

**Comment by naturalfreak** [comment]
Same problem when using Angular 9 and route resolver. NavigationEnd is emitted before initialization of the view, and scroll is restored on incorrect position. It's a serious bug after all this time with no update on this issue. But I think that it will be good to have another route event after initialization of the view because now I'm using asyncScheduler of rxjs module and it has negative effects.
Here is simple example for scroll restoration service:

```
import { Injectable } from "@angular/core";
import { Event, NavigationEnd, NavigationStart, Router } from '@angular/router';
import { ViewportScroller } from '@angular/common';
import { filter, observeOn, scan } from 'rxjs/operators';
import { asyncScheduler } from 'rxjs';

@Injectable({
  providedIn: 'root'
})

export class RouterScrollProvider {

  constructor(
    private router: Router,
    private viewportScroller: ViewportScroller
  ) { }

  setRouterListener() {
    this.router.events
      .pipe(
        filter(
          event => event instanceof NavigationStart || event instanceof NavigationEnd,
        ),
        scan<Event, ScrollPositionRestore>((scroll, event) => {
          const extras = this.router.getCurrentNavigation().extras;
          return {
            update: extras.state ? extras.state.update : true,
            event,
            positions: {
              ...scroll.positions,
              ...(event instanceof NavigationStart
                ? {
                  [event.id]: this.viewportScroller.getScrollPosition(),
                }
                : {}),
            },
            trigger:
              event instanceof NavigationStart
                ? event.navigationTrigger
                : scroll.trigger,
            idToRestore:
              (event instanceof NavigationStart &&
                event.restoredState &&
                event.restoredState.navigationId + 1) ||
              scroll.idToRestore,
          }
        }),
        filter(
          ({ event, trigger, update }) => event instanceof NavigationEnd && !!trigger && update,
        ),
        observeOn(asyncScheduler),
      )
      .subscribe(({ trigger, positions, idToRestore }) => {

        if (trigger === 'imperative') {
          this.viewportScroller.scrollToPosition([0, 0]);
        }

        if (trigger === 'popstate') {
          const position = positions[idToRestore] || [0, 0];
          this.viewportScroller.scrollToPosition(position);
        }

      });
  }

}

interface ScrollPositionRestore {
  update: boolean,
  event: Event;
  positions: { [K: number]: [number, number] };
  trigger: 'imperative' | 'popstate' | 'hashchange';
  idToRestore: number;
}
```

**Comment by DrDager** [comment]
I have **resolved** this issue by **implementing custom scroll restoration behavior**.

The **reason** of this behavior is that **page didn't already render** and scrollToPosition no have effect.

There are not a good hack with timeout but it works.

```
export class AppModule {
  constructor(router: Router, viewportScroller: ViewportScroller) {
    router.events.pipe(
      filter((e): e is Scroll => e instanceof Scroll)
    ).subscribe(e => {
      if (e.position) {
        // backward navigation
        setTimeout(() => {viewportScroller.scrollToPosition(e.position); }, 0);
      } else if (e.anchor) {
        // anchor navigation
        setTimeout(() => {viewportScroller.scrollToAnchor(e.anchor); }, 0);
      } else {
        // forward navigation
        setTimeout(() => {viewportScroller.scrollToPosition([0, 0]); }, 0);
      }
    });
  }
}
```

**Comment by danielehrhardt** [comment]
Same Issue here.

**Comment by patrickbussmann** [comment]
Here, too.
Maybe someone can create a pull request with the solution @DrDager @naturalfreak @dsebastien

**Comment by Splaktar** [comment]
> ```ts
>       ...
>       } else if (e.anchor) {
>         // anchor navigation
>         setTimeout(() => {viewportScroller.scrollToAnchor(e.anchor); }, 0);
>       } else {
>       ...
> ```

Thank you for this tip that the `Scroll` event contains the anchor `string`. Previously I was doing this, but I didn't like that it was disconnected from my `router.events` stream:
```ts
    route.fragment.subscribe((fragment: string) => (this.fragment = fragment));
```

As for the `setTimeout`s, my approach was to instead just add a small delay via `delay(100),` in the `router.events` stream.

As for the `Scroll.position` field, that's really helpful as well if you aren't using the `MatSidenav`. In that case, the `Scroll.position` is based on the `ViewportScroller` and not the `MatSidenavContent` scrollable, which means that its value is always `[0, 0]`. It looks like I'll have to dig into creating my own scroll position cache to handle this.

**Comment by hiepxanh** [comment]
Why this solution is hard to resolve like this?  It's almost 2 years

**Comment by Splaktar** [comment]
> I've tried using `scrollPositionRestoration` and indeed face the troubles mentioned in this issue. I've documented my findings and current solution here, just in case anyone is also looking for a solution: https://medium.com/@dSebastien/handling-scrolling-on-angular-router-transitions-e7652e57d964

This is a really helpful post! It also links to some other helpful posts on this topic. It would be great if someone on the team could evaluate all of these ideas/approaches and turn them into built in module, possibly with multiple strategies, to resolve the majority of scroll restoration and scroll positioning issues.

**Comment by LorisBachert** [comment]
I just stumbled upon this issue. In my case my scroll container unfortunately is not the window therefore the scrollPositionRestoration 'default' does not work. I wrote a quick workaround and just wanted to leave this here in hope it might help someone:
```typescript
export class AppRoutingModule {

  scrollTopPositions: { [url: string]: number } = {};

  constructor(router: Router) {
    router.events.pipe(
      filter((e: RouterEvent) => e instanceof NavigationStart || e instanceof NavigationEnd)
    ).subscribe({
      next: (e: RouterEvent) => {
        const scrollContainer = document.getElementById('scrollableContent');
        if (e instanceof NavigationStart) {
          this.scrollTopPositions[window.location.hash.substring(1)] = scrollContainer.scrollTop;
        } else if (e instanceof NavigationEnd) {
          const newUrl = e.url;
          setTimeout(() => {
            scrollContainer.scrollTo(0, this.scrollTopPositions[newUrl]);
          }, 0);
        }
      }
    });
  }
}
```

**Comment by epanchenko** [comment]
Thank you LorisBachert, I have a mat side nav, and your code solved my problem.

**Comment by kolombet** [comment]
I have custom container used for scrolling instead of scrolling page body. I created custom ViewPortScroller implementation, but was unable to inject in instead of BrowserViewportScroller. Is it possible to provide custom factory for ViewPortScroller provider?

**Comment by DeanPDX** [comment]
How hard would it be to provide some sort of config option to set your scrolling element? Because I'm running in to this same problem and traced my router and found that the events are firing as they should, but due to the Angular Material config I'm using, `body` isn't my scrolling element for my app (it's my content div, with body/html being height 100%).

One other easy option I can see would be to allow use of custom implementations of `ViewportScroller`. I'm looking in to that now and will update with stackblitz if I can easily get it working.

@LorisBachert I appreciate that example. I'm not sure if you are concerned about this, but, that code isn't 100% safe due to the fact that it creates a tight coupling between your app and rendering layers. Via [the docs](https://angular.io/api/core/ElementRef):

> Relying on direct DOM access creates tight coupling between your application and rendering layers which will make it impossible to separate the two and deploy your application into a web worker.

If you wanted to double-check that you're running in a browser, I've implemented that using a method like this:

```typescript
import { Inject, PLATFORM_ID } from '@angular/core';
import { isPlatformBrowser } from '@angular/common';

export class ExampleClass {
  /**
   * Indicates whether the current platform is a web browser and thus safe for DOM-specific code.
   */
  private isRunningInBrowser: boolean = null;

  constructor(@Inject(PLATFORM_ID) platformId: string) {
    this.isRunningInBrowser = isPlatformBrowser(platformId);
  }

  domSpecificFunction() {
    // Make sure we're running in a browser
    if (!this.isRunningInBrowser) {
      return;
    }
  }
}
```

Anyway, it probably doesn't matter in your environment but I figured I'd mention it.

**Comment by DeanPDX** [comment]
@kolombet I was able to get a custom ViewportScroller implementation working and it solved my problem. In your app module, use a factory just like Angular does internally with their [BrowserViewportScroller](https://github.com/angular/angular/blob/d1ea1f4c7f3358b730b0d94e65b00bc28cae279c/packages/common/src/viewport_scroller.ts#L27) implementation:

```typescript
providers: [
  {
    provide: ViewportScroller,
    useFactory: () => new CustomViewportScroller('content-scroller', ɵɵinject(DOCUMENT), window, ɵɵinject(ErrorHandler))
  },
```

In this simple implementation, I'm passing a string with an ID of my scrolling element (a div in this case). You would obviously want that to be in a config file or something. My custom constructor looks like this:

```typescript
constructor(private scrollElementID: string, private document: Document, private window: any, private errorHandler: ErrorHandler) { }
```

After that, I just modified `getScrollPosition` and `scrollToPosition` to be whatever I want. Something like this might work:

```typescript
/**
 * Retrieves the current scroll position.
 * @returns The position in screen coordinates.
 */
getScrollPosition(): [number, number] {
    const scrollEl = this.document.querySelector(`#${this.scrollElementID}`);
    if (this.supportScrollRestoration() && scrollEl) {
        return [scrollEl.scrollLeft, scrollEl.scrollTop];
    } else {
        return [0, 0];
    }
}

/**
 * Sets the scroll position.
 * @param position The new position in screen coordinates.
 */
scrollToPosition(position: [number, number]): void {
    const scrollEl = this.document.querySelector(`#${this.scrollElementID}`);
    if (this.supportScrollRestoration() && scrollEl) {
        // Total hack but waiting for for content/images to load to give us a 
        // better chance of hitting our scroll target. It also gives the UI a bit
        // of movement to show users that we scrolled them after page load. In a
        // real implementation of ViewportScroller, we should get rid of this but
        // it suits my current needs.
        setTimeout(() => {
            scrollEl.scrollTo(position[0], position[1]);    
        }, 200);
    }
}
```

Anyway, hopefully this helps anybody who wants to go with a custom `ViewportScroller` implementation. [Here's a working StackBlitz demonstrating all of this](https://stackblitz.com/edit/angular-custom-viewport-scroller?file=src%2Fapp%2Fcustom-viewport-scroller.ts). 

And final word of caution, as with my [previous comment about tightly coupling yourself to the DOM](https://github.com/angular/angular/issues/24547#issuecomment-648469322), you need to make sure you're executing on a browser.

**Comment by Vivida1** [comment]
Disabling 
``` css
scroll-behavior: smooth;
``` 
worked for me in FF.

**Comment by bibhas2** [comment]
A general approach used here by most is to delay the call to ``viewportScroller.scrollToPosition()``. This is not a very viable solution. If you are rendering data from a backend then it may take several seconds to fetch the data and fully populate the page. I think what we need is a simple API in the ``ActivatedRoute`` service to restore the scroll position. We should be able to call this API at any point to restore the scroll position.

**Comment by webme** [comment]
### For those who use Angular Material or Ionic ...
Here is how I solved the scroll problem.
For Ionic, it will be necessary to adapt the code, it is the same principle.

> In app.component.html add `#sidenavContent` reference 

```html
<mat-sidenav-container>
  ...
  <mat-sidenav-content #sidenavContent>
    ...
  </mat-sidenav-content>
</mat-sidenav-container>
```

> In app.component.ts

```javascript
@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.scss']
})
export class AppComponent implements OnInit, OnDestroy {
  @ViewChild('sidenavContent', {read: MatSidenavContent}) sidenavContentScrollable?: MatSidenavContent;

  subscriptions$ = new Subject<any>();

  constructor(
    @Inject(DOCUMENT) private document: Document,
    private router: Router
  ) {
  }

  ngOnInit(): void {
    this.router.events
      .pipe(
        filter((event: Event): event is Scroll => event instanceof Scroll),
        takeUntil(this.subscriptions$)
      )
      .subscribe((event: Scroll) => {
        if (!this.sidenavContentScrollable) {
          return;
        }

        const nativeElement = this.sidenavContentScrollable.getElementRef().nativeElement;

        if (event.position) {
          // backward navigation
          nativeElement.scroll(event.position[0], event.position[1]);
        } else if (event.anchor) {
          // anchor navigation
          const offsetTop = this.document.getElementById(event.anchor)?.offsetTop || 0;

          nativeElement.scroll(0, offsetTop);
        } else {
          // forward navigation
          nativeElement.scroll(0, 0);
          
          // An alternative the scrolling box scrolls in a smooth fashion, see browser compatibility before
          // https://developer.mozilla.org/en-US/docs/Web/API/Element/scroll#browser_compatibility

          // this.sidenavContentScrollable.scrollTo({left: 0, top: 0, behavior: 'smooth'});
        }
      });
  }

  ngOnDestroy(): void {
    this.subscriptions$.next();
    this.subscriptions$.complete();
  }
}
```

**Comment by IRCraziestTaxi** [comment]
I'm having trouble with scrolling to fragments while using animations as pointed out by @PapaNappa. Without animations it was working great, but with animations it scrolls to about halfway through the new view. Any suggestions on this issue?

**Comment by IRCraziestTaxi** [comment]
I actually managed to get the problem with fragments and animations working smoothly. I wasn't using the animations exactly right, but after taking (and modifying styles to my taste) code from [the docs](https://angular.io/guide/route-animations), it straightened up. I also had to tweak the container around my router outlet since I'm using UIKit's `uk-container`, which has padding in it, so with the absolute positioning on the animated views, it needed another div inside of the `uk-container` in which to put the router outlet. Hopefully this will help someone.

`routing.animations.ts`
```ts
import { animate, group, query, style, transition, trigger } from "@angular/animations";

export const routeChangeAnimation =
    trigger("routeAnimations", [
        transition("* => *", [
            style({
                position: "relative"
            }),
            query(
                ":enter, :leave",
                [
                    style({
                        position: "absolute",
                        top: 0,
                        left: 0,
                        width: "100%"
                    })
                ],
                {
                    optional: true
                }
            ),
            query(
                ":enter",
                [
                    style({
                        left: "100%",
                        opacity: 0
                    })
                ],
                {
                    optional: true
                }
            ),
            group([
                query(
                    ":leave",
                    [
                        animate("500ms ease-out", style({
                            left: "-100%",
                            opacity: 0
                        }))
                    ],
                    {
                        optional: true
                    }
                ),
                query(
                    ":enter",
                    [
                        animate("500ms ease-out", style({
                            left: 0,
                            opacity: 1
                        }))
                    ],
                    {
                        optional: true
                    }
                )
            ])
        ])
    ]);
```

`app.component.html`
```html
<!-- navbar and stuff. -->

<div class="uk-container uk-margin-bottom">
    <!-- id="lead" is my fragment for scrolling to the top of the page in router links. -->
    <div id="lead" [@routeAnimations]="prepareRoute(outlet)">
        <router-outlet #outlet="outlet"></router-outlet>
    </div>
</div>
```

`app.component.ts`
```ts
import { ViewportScroller } from "@angular/common";
import { Component } from "@angular/core";
import { RouterOutlet } from "@angular/router";
import { routeChangeAnimation } from "../animations/routing.animations";

@Component({
    selector: "app-root",
    templateUrl: "./app.component.html",
    styleUrls: ["./app.component.scss"],
    animations: [
        routeChangeAnimation
    ]
})
export class AppComponent {
    public constructor(
        private readonly _viewportScroller: ViewportScroller
    ) {
        // Offset y = height of uk-navbar.
        this._viewportScroller.setOffset([0, 80]);
    }

    public prepareRoute(outlet: RouterOutlet): string {
        return outlet?.activatedRouteData?.animation;
    }
}
```

Jumping to other fragments (not at the top of a page) also scrolls to the correct point in the page. My only problem now is that when a router link scrolls to the bottom of the page in the target view, the margin from the `uk-margin-bottom` I have in `app.component.html` doesn't take effect until the view finishes animating in, causing a mild jump of the page content. Not exactly sure why, but it's a lot better than it was.

**Comment by JulMack** [comment]
A nice option would be to allow to set a custom reference for the scroll-behavior of the RouterModule to attach to. So that the call signature would be like 
```
imports: [RouterModule.forRoot(routes, {
  scrollPositionRestoration: 'enabled', 
  scrollElementReference: 'scrollMe'
})]
```
with a sidenav for example:
```
<mat-sidenav-container>
  ...
  <mat-sidenav-content #scrollMe>
    ...
  </mat-sidenav-content>
</mat-sidenav-container>
```

**Comment by hiepxanh** [comment]
we should remove `filter((e:Event) =>` to ` filter((e) =>` to prevent type checking error.
and I have to change scroll to have setTimeout:
```
setTimeout(() => {
          e.position && this.viewportScroller.scrollToPosition(e.position);
        }, 300);
``` 
to prevent scroll trigger too fast, before ngOnInit. I dont know why my component both reconstruct and re run ngOnInit on back. I just back only. and it cannot run fast enough

```
scrollRestoration() {
    // https://angular.io/api/router/ExtraOptions#scrollPositionRestoration
    this.router.events.pipe(filter((e): e is Scroll => e instanceof Scroll)).subscribe((e) => {
      if (e.position) {
        // console.log('e.position scrolling....', e);
        // backward navigation
        setTimeout(() => {
          e.position && this.viewportScroller.scrollToPosition(e.position);
        }, 300);
      } else if (e.anchor) {
        // console.log('e.position anchor!!', e);
        // anchor navigation
        this.viewportScroller.scrollToAnchor(e.anchor);
      } else {
        // forward navigation
        // console.log('e.position scroll top on new navigation', e);
        this.viewportScroller.scrollToPosition([0, 0]);
      }
    });
```

**Comment by TheJLifeX** [comment]
Hi everybody,

concerning this issue, you could have a look at [NgxScrollPositionRestoration](https://github.com/TheJLifeX/ngx-scroll-position-restoration). A library for scroll position restoration in Angular.

The library supports scroll position restoration on:
- Any scrollable element (for example: `.mat-sidenav-content` in case you are using [Angular Material Sidenav](https://material.angular.io/components/sidenav/overview))
- Lazy loading content
- Named router-outlets (multiple router-outlets)
- Child routes (nested router-outlets)
- Backward and forward navigation

## [View the demo](https://thejlifex.github.io/ngx-scroll-position-restoration)

**Comment by HARSHITHASAI-A** [comment]
Hello everyone,
 
I have the similar issue. I have used mat-paginator for pagination and on the new/next page click the scrollbar should reset to top for the page change event. Instead it is shortening, could any one please help me with this. @TheJLifeX  @hiepxanh 

.html file

`<div class="content hubble-scrollbar">
  <div class="result-container"
       *ngFor="let result of results?.data"
       [ngClass]="{
      selected: selectedItemId === result.id
    }">
    <div>
      <hubble-document-content-annotation [document]="result"></hubble-document-content-annotation>
    </div>
    <span class="date mat-caption">{{
      result.f_received | date: "medium"
      }}</span>
    <div class="title-container">
      <a class="title"
         [routerLink]='"."'
         [fragment]="result.id"
         (click)="onSelect(result.id, result.datasetId)">{{ result.f_subject }}</a>
      <a class="parent-link"
         *ngIf="result.f_thread_id">
        <mat-icon color="primary">dynamic_feed</mat-icon>
      </a>
      <a class="parent-link"
         *ngIf="result.f_attachment_count>0">
        <mat-icon color="primary">attach_file</mat-icon>
      </a>
      <mat-chip-list *ngIf="priorityKey"
                     class="priority">
        <mat-chip class="small"
                  [matTooltip]="priorityKey | priorityName | async"
                  [style.backgroundColor]="result | priorityColor: priorityKey">{{ result | priorityLabel: priorityKey
          }}</mat-chip>
      </mat-chip-list>
    </div>
    <div class="snippet truncate mat-body-2">{{ result.snippet }}</div>
    <span class="body-content"
          [innerHTML]="result.f_body">
    </span>
  </div>
  <div class="paginator-container">
    <mat-paginator showFirstLastButtons
                   [length]="paginationDetails?.total || 0"
                   [pageSize]="50"
                   hidePageSize="true"
                   [hubblePaginator]="paginationDetails"
                   (page)="page($event)"></mat-paginator>
  </div>
</div>
`

.ts file - page method

`page(event: PageEvent) {
    this.pagination.page = event.pageIndex + 1;
    this.pagination.pageSize = event.pageSize;
    this.changePage.emit(this.pagination);
    this.router.navigate([], {
      relativeTo: this.activatedRoute,
      queryParams: { pageIndex: event.pageIndex + 1, pageSize: event.pageSize },
      queryParamsHandling: 'merge'
    });
  }`

**Comment by TheJLifeX** [comment]
Hi @HARSHITHASAI-A,
Since you are using a paginator, you should know that scroll position restauration only happens on popstate navigation (when you click on the browser back- or forward button). When you click on the paginator-buttons (even when you return to the previous page) you create an imperative navigation (with `this.router.navigate(…)`).
To manually trigger a popstate navigation you can use the following functions:
```ts
import { Location } from '@angular/common';

class SomeComponent {

  constructor(private location: Location) { }

  goBack() {
    this.location.back(); // See: https://angular.io/api/common/Location#back
  }

  goForward() {
    this.location.forward(); // See: https://angular.io/api/common/Location#forward
  }
}
```
___
If you are using [NgxScrollPositionRestoration](https://www.npmjs.com/package/ngx-scroll-position-restoration), you can import it like this (with `debug` set to `true`), to debug if something is not working as expected:
```ts
    NgxScrollPositionRestorationModule.forRoot({
      debug: true
    })
```

I hope this can help you.

**Comment by Erbenos** [comment]
It seems that af8afee addresses some change detection portion of this issue, though I am not if that anyhow resolves the issue for those people that don't have html/body their scrollable element. It would be nice if we could pass either a specific selector for scrollable container ourselves or some kind of custom strategy for resolving the component dynamically ourselves.

**Comment by Goretzky** [comment]
I have a simple Angular 14 app where everything is served prerendered on S3. Angular's scrollPositionRestoration: 'enabled' works for my web app as long as the screen is large. But on a mobile phone screen scrollPositionRestoraiton stops working. Anyone know why this would happen? 

I ran into problems trying DeanPDX's solution, which has an Angular 9 demo. For example, I wasn't able to import eeinject in the app.module.ts. Was anyone able to get this solution working in Angular 14? If so, what changes did you have to make?

NgxScrollPositionRestoration is only compatible with Angular 13 and earlier, or did someone get this to work in Angular 14?

Small screen size issue happened in Chrome browser on a large monitor with browser shrunk down, on a chromebook again with the browser shrunk down to mobile phone size, and an iphone using Chrome browser and Safari. I also tried anchorScrolling as a solution but had the same result: works on wide screen but not on cell phone sized screen.
